docs/source/asr/datasets.rst

```diff
@@ -823,6 +823,52 @@ For multi-dataset setups, one may provide multiple manifests and even their weig
             bucket_duration_bins=[1.91,3.02,3.56,...
     <other diagnostic information about the dataset>
 
+
+Seeds and randomness
+~~~~~~~~~~~~~~~~~~~~
+
+In Lhotse dataloading configuration we have two parameters controlling randomness: ``seed`` and ``shard_seed``.
+Both of them can be either set to a fixed number, or one of two string options ``"randomized"`` and ``"trng"``.
+Their roles are:
+
+    * ``seed`` is the base random seed, and is one of several factors used to initialize various RNGs participating in dataloading.
+
+    * ``shard_seed`` controls the shard randomization strategy in distributed data parallel setups when using sharded tarred datasets.
+
+Below are the typical examples of configuration with an explanation of the expected outcome.
+
+Case 1 (default): ``seed=<int>`` and ``shard_seed="trng"``:
+
+    * The ``trng`` setting discards ``seed`` and causes the actual random seed to be drawn using OS's true RNG. Each node/GPU/dataloading worker draws its own unique random seed when it first needs it.
+
+    * Each node/GPU/dataloading worker yields data in a different order (no mini-batch duplication).
+
+    * On each training script run, the order of dataloader examples are **different**.
+
+    * Since the random seed is unpredictable, the exact dataloading order is not replicable.
+
+Case 2: ``seed=<int>`` and ``shard_seed="randomized"``:
+
+    * The ``randomized`` setting uses ``seed`` along with DDP ``rank`` and dataloading ``worker_id`` to set a unique but deterministic random seed in each dataloading process across all GPUs.
+
+    * Each node/GPU/dataloading worker yields data in a different order (no mini-batch duplication).
+
+    * On each training script run, the order of dataloader examples are **identical** as long as ``seed`` is the same.
+
+    * This setup guarantees 100% dataloading reproducibility.
+
+    * Resuming training without changing of the ``seed`` value will cause the model to train on data it has already seen. For large data setups, not managing the ``seed`` may cause the model to never be trained on a majority of data. This is why this mode is not the default.
```

Comment:
what's the recommended practice on managing the "seed" on large data setup?

Reply:
Generally every time you resume, you'd provide a different value to `model.train_ds.seed=<val>`. A true-enough random seed can be obtained on most systems by reading /dev/urandom, e.g. uint32 seed: `RSEED=$(od -An -N4 -tu4 < /dev/urandom | tr -d ' ')`. If you have some sort of "launcher script" that queues multiple jobs, this would be the right place to use this.

Ideally we'd be able to automate this seed management thing by keeping some state in the checkpoints, but at this point it'd be a scope creep.